src/System.Net.Http/src/System/Net/Http/Unix/CurlHandler.cs

```diff
@@ -415,15 +502,231 @@ private void CheckDisposedOrStarted()
+            if (result != CURLcode.CURLE_OK)
+            {
+                throw new HttpRequestException(SR.net_http_client_execution_error, GetCurlException(result));
+            }
+        }
+
+        private void SetCurlMultiOptions()
+        {
+            _multiHandlePtr = GCHandle.Alloc(_multiHandle);
+            IntPtr callbackContext = GCHandle.ToIntPtr(_multiHandlePtr);
+            int result = Interop.libcurl.curl_multi_setopt(_multiHandle, CURLMoption.CURLMOPT_SOCKETFUNCTION,
+                (Interop.libcurl.curl_socket_callback) CurlSocketCallback);
+            if (result == CURLMcode.CURLM_OK)
+            {
+                result = Interop.libcurl.curl_multi_setopt(_multiHandle, CURLMoption.CURLMOPT_TIMERFUNCTION,
+                    (Interop.libcurl.curl_multi_timer_callback) CurlMultiTimerCallback);
+            }
+            if (result == CURLMcode.CURLM_OK)
+            {
+                result = Interop.libcurl.curl_multi_setopt(_multiHandle, CURLMoption.CURLMOPT_TIMERDATA, callbackContext);
+            }
+            if (result != CURLMcode.CURLM_OK)
+            {
+                throw new HttpRequestException(SR.net_http_client_execution_error, GetCurlException(result, true));
+            }
+        }
+
+        private SafeCurlSlistHandle SetRequestHeaders(SafeCurlHandle handle, HttpRequestMessage request)
+        {
+            SafeCurlSlistHandle retVal = new SafeCurlSlistHandle();
+            if (request.Headers == null)
+            {
+                return retVal;
+            }
+
+            HttpHeaders contentHeaders = null;
+            if (request.Content != null)
+            {
+                // TODO: Content-Length header isn't getting correctly placed using ToString()
+                // This is a bug in HttpContentHeaders that needs to be fixed.
+                if (request.Content.Headers.ContentLength.HasValue)
+                {
+                    long contentLength = request.Content.Headers.ContentLength.Value;
+                    request.Content.Headers.ContentLength = null;
+                    request.Content.Headers.ContentLength = contentLength;
+                }
+                contentHeaders = request.Content.Headers;
+            }
+
+            string[] allHeaders = HeaderUtilities.DumpHeaders(request.Headers, contentHeaders)
+                .Split(s_headerDelimiters, StringSplitOptions.RemoveEmptyEntries);
+            bool gotReference = false;
+            try
+            {
+                for (int i = 0; i < allHeaders.Length; i++)
+                {
+                    string header = allHeaders[i];
+                    if (header.Equals("{") || header.Equals("}"))
+                    {
+                        continue;
+                    }
+                    IntPtr rawHandle = IntPtr.Zero;
+                    if (!retVal.IsInvalid)
+                    {
+                        retVal.DangerousAddRef(ref gotReference);
```

Comment:
This logic doesn't look right.  Seems like we're going to potentially AddRef length times, but only ReleaseRef once?